cmd/src/campaigns_apply.go

```diff
@@ -0,0 +1,233 @@
+		fileFlag             = flagSet.String("f", "", "The campaign spec file to read.")
+		keepFlag             = flagSet.Bool("keep-logs", false, "Retain logs after executing steps.")
+		namespaceFlag        = flagSet.String("namespace", "", "The user or organization namespace to place the campaign within.")
+		parallelismFlag      = flagSet.Int("j", 0, "The maximum number of parallel jobs. (Default: GOMAXPROCS.)")
+		timeoutFlag          = flagSet.Duration("timeout", 60*time.Minute, "The maximum duration a single set of campaign steps can take.")
+		apiFlags             = api.NewFlags(flagSet)
+	)
+
+	var (
+		pendingColor = output.StylePending
+		successColor = output.StyleSuccess
+		successEmoji = output.EmojiSuccess
+	)
+
+	createPending := func(out *output.Output, message string) output.Pending {
+		return out.Pending(output.Line("", pendingColor, message))
+	}
+
+	completePending := func(p output.Pending, message string) {
+		p.Complete(output.Line(successEmoji, successColor, message))
+	}
+
+	doApply := func(out *output.Output) error {
+		ctx := context.Background()
+		client := cfg.apiClient(apiFlags, flagSet.Output())
+
+		// Parse flags and build up our service options.
+		var errs *multierror.Error
+		svc := campaigns.NewService(&campaigns.ServiceOpts{
+			AllowUnsupported: *allowUnsupportedFlag,
+			Client:           client,
+		})
+
+		specFile, err := campaignsOpenFileFlag(fileFlag)
+		if err != nil {
+			errs = multierror.Append(errs, err)
+		} else {
+			defer specFile.Close()
+		}
+
+		if namespaceFlag == nil || *namespaceFlag == "" {
+			errs = multierror.Append(errs, &usageError{errors.New("a namespace must be provided with -namespace")})
+		}
+
+		opts := campaigns.ExecutorOpts{
+			Cache:      svc.NewExecutionCache(*cacheDirFlag),
+			ClearCache: *clearCacheFlag,
+			KeepLogs:   *keepFlag,
+			Timeout:    *timeoutFlag,
+		}
+		if parallelismFlag != nil || *parallelismFlag <= 0 {
+			opts.Parallelism = runtime.GOMAXPROCS(0)
+		} else {
+			opts.Parallelism = *parallelismFlag
+		}
+		executor := svc.NewExecutor(opts, nil)
+
+		if errs != nil {
+			return errs
+		}
+
+		pending := createPending(out, "Parsing campaign spec")
+		campaignSpec, err := svc.ParseCampaignSpec(specFile)
+		if err != nil {
+			return errors.Wrap(err, "parsing campaign spec")
+		}
+
+		if err := campaignsValidateSpec(out, campaignSpec); err != nil {
+			return err
+		}
+		completePending(pending, "Parsing campaign spec")
+
+		pending = createPending(out, "Resolving namespace")
+		namespace, err := svc.ResolveNamespace(ctx, *namespaceFlag)
+		if err != nil {
+			return err
+		}
+		completePending(pending, "Resolving namespace")
+
+		var progress output.Progress
+		specs, err := svc.ExecuteCampaignSpec(ctx, executor, campaignSpec, func(statuses []*campaigns.TaskStatus) {
+			if progress == nil {
+				progress = out.Progress([]output.ProgressBar{{
+					Label: "Executing campaign spec",
```

Comment:
I don't think executing _the spec_ makes sense. Instead, maybe something like this:

```suggestion
					Label: "Executing steps",
```